File: go/internal/messaging/streams.go
package messaging

import (
	"context"
	"time"

	"github.com/nats-io/nats.go"
)

const (
	StreamName = "EVENTS"
	// EventSchemaVersion is stamped on every published event; bump on a breaking
	// envelope change. Consumers reject versions they do not support.
	EventSchemaVersion = 1
)

func EnsureStreams(ctx context.Context, js nats.JetStreamContext) error {
	// Create stream if missing (idempotent)
	_, err := js.StreamInfo(StreamName)
	if err == nil {
		return nil
	}
	_, err = js.AddStream(&nats.StreamConfig{
		Name:              StreamName,
		Subjects:          []string{"events.>"},
		Storage:           nats.FileStorage,
		Retention:         nats.LimitsPolicy,
		MaxMsgsPerSubject: 1000000,
		Discard:           nats.DiscardOld,
		Duplicates:        2 * time.Minute, // 2 minutes
	})
	return err
}

File: rust/sim/src/messaging/streams.rs
use async_nats::jetstream;
use std::time::Duration;

pub const STREAM_NAME: &str = "EVENTS";

pub async fn ensure_streams(js: &jetstream::Context) -> Result<(), async_nats::Error> {
    js.get_or_create_stream(jetstream::stream::Config {
        name: STREAM_NAME.into(),
        subjects: vec!["events.>".into()],
        storage: jetstream::stream::StorageType::File,
        retention: jetstream::stream::RetentionPolicy::Limits,
        max_messages_per_subject: 1_000_000,
        discard: jetstream::stream::DiscardPolicy::Old,
        duplicate_window: Duration::from_secs(120),
        ..Default::default()
    })
    .await?;
    Ok(())
}
